### ISSUE TITLE
static_assert cannot prevent OOM caused by numeric overflow

### ISSUE THREAD

**Issue by ingomueller-net** [llvm/llvm-project#54144]
Consider the following snippet:

```cpp
// Compile with: clang-15 -std=c++14 test.cpp
#include <utility>

int main(int, char**) {
  static constexpr std::size_t kA = 2;
  static constexpr std::size_t kB = 1;
  static_assert(kA <= kB, "Should fail");                     // Prints an error on std::out but compilation continues
  static constexpr std::size_t kMinusOne = kB - kA;           // Produces an overflow
  using IndexSequence = std::make_index_sequence<kMinusOne>;  // Makes clang++ run out of memory
}
```

The `static_assert` fails; however, compilation continues, presumably for diagnostics. The subsequent code produces a numeric overflow, which causes a huge type to be created (namely `std::make_index_sequence<18446744073709551615>`), which makes the compiler run until it runs out of memory.

The problem is two-fold:
1. The compiler should not run out of memory but give up before that. In many setups, continuing until the kernel kills the compiler may freeze the machine.
2. It does not seem possible to prevent the compiler from doing this mistake since it continues to compile even after the failed assert. (Admittedly, diagnostics after the first error are helpful, so not sure if that has a fix.)

**Comment by dwblaikie** [comment]
Yeah, I don't think (2) will be addressable, but (1) should be. Clang has a bunch of recursion limits - but I guess this particular case is maybe slipping through somewhere/somehow.

Hopefully @AaronBallman and @zygoloid have some more idea where the missing recursion limiter is.

**Comment by AaronBallman** [comment]
I haven't had the chance to investigate much, but I'm not certain this issue is in the Clang frontend. I think it could be in libc++ instead. Here's a different example using the compiler intrinsic instead of the STL function, and it diagnoses properly: https://godbolt.org/z/j55dj1zY7.

**Comment by ingomueller-net** [comment]
I am by no means an authority in this questions but I'd say that the compiler should give up eagerly before running OoM no matter the input, i.e., the problem is at least *also* in the compiler. It may be a good idea to add a check/constraint/... into libc++ to have a more precise/early warning, but not as the only fix.

**Comment by AaronBallman** [comment]
> I am by no means an authority in this questions but I'd say that the compiler should give up eagerly before running OoM no matter the input, i.e., the problem is at least _also_ in the compiler. It may be a good idea to add a check/constraint/... into libc++ to have a more precise/early warning, but not as the only fix.

That's a fair way of approaching it as well; we would usually hit a template instantiation depth limit far before ever hitting an OOM situation, so I think this deserves a look from the frontend as well as libc++.

**Comment by ivafanas** [comment]
Compiler problem is triggered in `_make_integer_sequence` clang builtin. Compiler frontend continues pushing back new arguments to template until OOM happened.

The fast and very simple fix for the issue would be to provide command line option to limit `_make_integer_sequence` clang builtin arguments size. But it would not fixing the generic problem, imho.

I assume that nice solution would be to provide command line option which limits template arguments count for all templates instantiations. It is not so simple as `-ftemplate-depth` option because arguments list is already created (and OOM happened) before template instantiation call where `-ftemplate-depth` check comes to play.
